File: apps/java/src/test/java/algorithm/SearchTest.java
package algorithm;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class SearchTest {

    @Nested
    class LinearSearchTest {
        @Test
        void while文で探索_見つかる() {
            assertEquals(3, Search.linearSearchWhile(new int[]{6, 4, 3, 2, 1, 2, 8}, 2));
        }

        @Test
        void while文で探索_見つからない() {
            assertEquals(-1, Search.linearSearchWhile(new int[]{1, 2, 3}, 99));
        }

        @Test
        void for文で探索_見つかる() {
            assertEquals(3, Search.linearSearchFor(new int[]{6, 4, 3, 2, 1, 2, 8}, 2));
        }

        @Test
        void for文で探索_見つからない() {
            assertEquals(-1, Search.linearSearchFor(new int[]{1, 2, 3}, 99));
        }

        @Test
        void 番兵法で探索_見つかる() {
            assertEquals(3, Search.linearSearchSentinel(new int[]{6, 4, 3, 2, 1, 2, 8}, 2));
        }

        @Test
        void 番兵法で探索_見つからない() {
            assertEquals(-1, Search.linearSearchSentinel(new int[]{1, 2, 3}, 99));
        }
    }

    @Nested
    class BinarySearchTest {
        @Test
        void 中央の要素を探索() {
            assertEquals(3, Search.binarySearch(new int[]{1, 2, 3, 5, 7, 8, 9}, 5));
        }

        @Test
        void 先頭の要素を探索() {
            assertEquals(0, Search.binarySearch(new int[]{1, 2, 3, 5, 7, 8, 9}, 1));
        }

        @Test
        void 末尾の要素を探索() {
            assertEquals(6, Search.binarySearch(new int[]{1, 2, 3, 5, 7, 8, 9}, 9));
        }

        @Test
        void 見つからない() {
            assertEquals(-1, Search.binarySearch(new int[]{1, 2, 3, 5, 7, 8, 9}, 4));
        }
    }

    @Nested
    class ChainedHashTest {
        private Search.ChainedHash h;

        @BeforeEach
        void setUp() {
            h = new Search.ChainedHash(13);
            h.add(1, "赤尾");
            h.add(5, "武田");
            h.add(10, "小野");
            h.add(12, "鈴木");
            h.add(14, "神崎");
        }

        @Test
        void 探索_見つかる() {
            assertEquals("赤尾", h.search(1));
            assertEquals("神崎", h.search(14));
        }

        @Test
        void 探索_見つからない() {
            assertNull(h.search(100));
        }

        @Test
        void 追加して探索() {
            h.add(100, "山田");
            assertEquals("山田", h.search(100));
        }

        @Test
        void 重複キーは追加できない() {
            assertFalse(h.add(1, "重複"));
        }

        @Test
        void 削除() {
            h.add(100, "山田");
            assertTrue(h.remove(100));
            assertNull(h.search(100));
        }

        @Test
        void 衝突したノードの削除() {
            assertTrue(h.remove(1));
            assertTrue(h.remove(14));
            assertNull(h.search(1));
            assertNull(h.search(14));
        }

        @Test
        void 存在しないキーの削除() {
            assertFalse(h.remove(999));
        }
    }

    @Nested
    class OpenHashTest {
        private Search.OpenHash h;

        @BeforeEach
        void setUp() {
            h = new Search.OpenHash(13);
            h.add(1, "赤尾");
            h.add(5, "武田");
            h.add(10, "小野");
            h.add(12, "鈴木");
            h.add(14, "神崎");
        }

        @Test
        void 探索_見つかる() {
            assertEquals("赤尾", h.search(1));
        }

        @Test
        void 探索_見つからない() {
            assertNull(h.search(999));
        }

        @Test
        void 追加して探索() {
            h.add(100, "山田");
            assertEquals("山田", h.search(100));
        }

        @Test
        void 重複キーは追加できない() {
            assertFalse(h.add(1, "重複"));
        }

        @Test
        void 削除() {
            h.add(100, "山田");
            assertTrue(h.remove(100));
            assertNull(h.search(100));
        }

        @Test
        void 存在しないキーの削除() {
            assertFalse(h.remove(999));
        }

        @Test
        void テーブル満杯時の追加() {
            Search.OpenHash small = new Search.OpenHash(3);
            small.add(0, "a");
            small.add(1, "b");
            small.add(2, "c");
            assertFalse(small.add(99, "d"));
        }

        @Test
        void テーブル満杯時の存在しないキー削除() {
            Search.OpenHash small = new Search.OpenHash(3);
            small.add(0, "a");
            small.add(1, "b");
            small.add(2, "c");
            assertFalse(small.remove(99));
        }
    }
}

File: apps/csharp/Algorithm.Tests/SearchTest.cs
using Xunit;
using Algorithm;

namespace Algorithm.Tests;

/// <summary>第3章 探索アルゴリズム テスト</summary>
public class SearchTest
{
    public class LinearSearchTest
    {
        [Fact] public void while文で探索_見つかる() => Assert.Equal(3, Search.LinearSearchWhile([6, 4, 3, 2, 1, 2, 8], 2));
        [Fact] public void while文で探索_見つからない() => Assert.Equal(-1, Search.LinearSearchWhile([1, 2, 3], 99));
        [Fact] public void for文で探索_見つかる() => Assert.Equal(3, Search.LinearSearchFor([6, 4, 3, 2, 1, 2, 8], 2));
        [Fact] public void for文で探索_見つからない() => Assert.Equal(-1, Search.LinearSearchFor([1, 2, 3], 99));
        [Fact] public void 番兵法で探索_見つかる() => Assert.Equal(3, Search.LinearSearchSentinel([6, 4, 3, 2, 1, 2, 8], 2));
        [Fact] public void 番兵法で探索_見つからない() => Assert.Equal(-1, Search.LinearSearchSentinel([1, 2, 3], 99));
    }

    public class BinarySearchTest
    {
        [Fact] public void 中央の要素を探索() => Assert.Equal(3, Search.BinarySearch([1, 2, 3, 5, 7, 8, 9], 5));
        [Fact] public void 先頭の要素を探索() => Assert.Equal(0, Search.BinarySearch([1, 2, 3, 5, 7, 8, 9], 1));
        [Fact] public void 末尾の要素を探索() => Assert.Equal(6, Search.BinarySearch([1, 2, 3, 5, 7, 8, 9], 9));
        [Fact] public void 見つからない() => Assert.Equal(-1, Search.BinarySearch([1, 2, 3, 5, 7, 8, 9], 4));
    }

    public class ChainedHashTest
    {
        private Search.ChainedHash MakeHash()
        {
            var h = new Search.ChainedHash(13);
            h.Add(1, "赤尾"); h.Add(5, "武田"); h.Add(10, "小野");
            h.Add(12, "鈴木"); h.Add(14, "神崎");
            return h;
        }

        [Fact] public void 探索_見つかる() { var h = MakeHash(); Assert.Equal("赤尾", h.Search(1)); Assert.Equal("神崎", h.Search(14)); }
        [Fact] public void 探索_見つからない() => Assert.Null(MakeHash().Search(100));
        [Fact] public void 追加して探索() { var h = MakeHash(); h.Add(100, "山田"); Assert.Equal("山田", h.Search(100)); }
        [Fact] public void 重複キーは追加できない() => Assert.False(MakeHash().Add(1, "重複"));
        [Fact] public void 削除() { var h = MakeHash(); h.Add(100, "山田"); Assert.True(h.Remove(100)); Assert.Null(h.Search(100)); }
        [Fact] public void 衝突したノードの削除() { var h = MakeHash(); Assert.True(h.Remove(1)); Assert.True(h.Remove(14)); Assert.Null(h.Search(1)); Assert.Null(h.Search(14)); }
        [Fact] public void 存在しないキーの削除() => Assert.False(MakeHash().Remove(999));
    }

    public class OpenHashTest
    {
        private Search.OpenHash MakeHash()
        {
            var h = new Search.OpenHash(13);
            h.Add(1, "赤尾"); h.Add(5, "武田"); h.Add(10, "小野");
            h.Add(12, "鈴木"); h.Add(14, "神崎");
            return h;
        }

        [Fact] public void 探索_見つかる() => Assert.Equal("赤尾", MakeHash().Search(1));
        [Fact] public void 探索_見つからない() => Assert.Null(MakeHash().Search(999));
        [Fact] public void 追加して探索() { var h = MakeHash(); h.Add(100, "山田"); Assert.Equal("山田", h.Search(100)); }
        [Fact] public void 重複キーは追加できない() => Assert.False(MakeHash().Add(1, "重複"));
        [Fact] public void 削除() { var h = MakeHash(); h.Add(100, "山田"); Assert.True(h.Remove(100)); Assert.Null(h.Search(100)); }
        [Fact] public void 存在しないキーの削除() => Assert.False(MakeHash().Remove(999));

        [Fact]
        public void テーブル満杯時の追加()
        {
            var small = new Search.OpenHash(3);
            small.Add(0, "a"); small.Add(1, "b"); small.Add(2, "c");
            Assert.False(small.Add(99, "d"));
        }

        [Fact]
        public void テーブル満杯時の存在しないキー削除()
        {
            var small = new Search.OpenHash(3);
            small.Add(0, "a"); small.Add(1, "b"); small.Add(2, "c");
            Assert.False(small.Remove(99));
        }
    }
}
